### PR TITLE
Add launch stack service logic

### PR DESCRIPTION
**Pull request by dragorosson**

### CODE REVIEW

**Comment by stretchservers** [comment]
Can one of the admins verify this patch?

**Comment by stretchservers** [comment]
Can one of the admins verify this patch?

File: otter/convergence/service.py [hunk 1]
```diff
@@ -336,6 +342,10 @@ def execute_convergence(tenant_id, group_id, build_timeout, waiting,
     yield do_return(result)
 
 
+def update_stacks_cache(scaling_group, now, stacks, include_deleted=True):
+    return Effect(Func(lambda: None))
```
Comment (radix) [inline otter/convergence/service.py]: You can also use `Constant(None)` here, though it matters very little :)

File: otter/test/convergence/test_service.py [hunk 2]
```diff
@@ -1366,6 +1372,43 @@ def plan(*args, **kwargs):
...
+
+        def plan(*args, **kwargs):
+            return []
+
+        executor = attr.assoc(launch_stack_executor,
+                              gather=intent_func('gacd'),
+                              plan=plan)
+
+        seq = [
+            parallel_sequence([]),
```
Comment (radix) [inline otter/test/convergence/test_service.py]: Since these entries are kind of confusing I like having a comment here explaining why it's there (I know this isn't consistently done elsewhere in the file).  I think this is the here because we have no steps, right?

**Comment by radix** [comment]
Just minor points. Looking very good. I'm glad how easily this is getting "zipped up" at the end :)

File: otter/test/convergence/test_service.py (hunk 2 above)
Comment (dragorosson) [inline otter/test/convergence/test_service.py]: I added a comment for it. Yes, it was because there were no steps, but it was for logging the steps.

File: otter/convergence/service.py (hunk 1 above)
Comment (dragorosson) [inline otter/convergence/service.py]: I left this as is, but thanks for the tip :)

**Comment by dragorosson** [comment]
Added tests for get_executor as well.

File: otter/test/convergence/test_service.py [hunk 3]
```diff
@@ -1366,6 +1374,44 @@ def plan(*args, **kwargs):
...
+                                        pmap())),
+             dispatch(reference_dispatcher)),
+        ]
+        eff = (
+            execute_convergence(
+                self.tenant_id, self.group_id, build_timeout=3600,
+                waiting=self.waiting,
+                limited_retry_iterations=43,
+                get_executor=lambda _: executor)
+        )
```
Comment (manishtomar) [inline otter/test/convergence/test_service.py]: Non-blocking nitpick: Hmm.. I haven't noticed this kind of indentation before. Just curious why you didn't go with:
```python
eff = execute_convergence(
    self.tenant_id, self.group_id, build_timeout=3600, waiting=self.waiting,
    limited_retry_iterations=43, get_executor=lambda _: executor)
```

Comment (dragorosson) [inline otter/test/convergence/test_service.py]: Me neither, fixed!

File: otter/test/convergence/test_service.py [hunk 4]
```diff
@@ -1366,6 +1374,41 @@ def plan(*args, **kwargs):
             perform_sequence(self.get_seq() + sequence, self._invoke(plan)),
             ConvergenceIterationStatus.Stop())
 
+    def test_launch_stack_config(self):
+        """Using a launch_stack launch config works."""
```
Comment (manishtomar) [inline otter/test/convergence/test_service.py]: This test checks launch_stack config with no steps. It will be nice to mention that in the docstring. Also, we may need to one more test case which returns steps that checks if they are executed. That can be another PR though :)

File: otter/test/convergence/test_service.py [hunk 5]
```diff
@@ -1366,6 +1374,41 @@ def plan(*args, **kwargs):
...
+        seq = [
+            parallel_sequence([]),  # No steps to log (log_steps)
+            (Log(msg='execute-convergence', fields=mock.ANY), noop),
+            (Log(msg='execute-convergence-results', fields=mock.ANY), noop),
+            (ModifyReference(self.waiting,
+                             match_func(pmap({self.group_id: 43}),
+                                        pmap())),
+             dispatch(reference_dispatcher)),
+        ]
+        eff = execute_convergence(
```
Comment (manishtomar) [inline otter/test/convergence/test_service.py]: Non-blocking: There is `_invoke` function in the class to call this. Would it be possible to modify that to work with this test? One thing I can think of is `_invoke` taking `executor` as an argument that if not provided defaults to launch_server executor.

File: otter/test/convergence/test_service.py [hunk 6]
```diff
@@ -1366,6 +1374,41 @@ def plan(*args, **kwargs):
...
+            self.tenant_id, self.group_id, build_timeout=3600,
+            waiting=self.waiting, limited_retry_iterations=43,
+            get_executor=lambda _: executor)
+        perform_sequence(self.get_seq(with_cache=False) + seq, eff)
```
Comment (manishtomar) [inline otter/test/convergence/test_service.py]: Would be nice to ensure that return value is `ConvergenceIterationStatus.Stop()`.

**Comment by manishtomar** [comment]
LGTM modulo comments :)

File: otter/test/convergence/test_service.py (hunk 5 above)
Comment (dragorosson) [inline otter/test/convergence/test_service.py]: That works, good idea. Done.

File: otter/test/convergence/test_service.py (hunk 6 above)
Comment (dragorosson) [inline otter/test/convergence/test_service.py]: Done.

**Comment by manishtomar** [comment]
Jenkins please run integration tests

**Comment by manishtomar** [comment]
Please merge once integration tests pass :)

**Comment by dragorosson** [comment]
Jenkins please run integration tests

**Comment by radix** [comment]
jenkins please run integration tests

**Comment by dragorosson** [comment]
jenkins please run integration tests

**Comment by radix** [comment]
jenkins please run integration tests
